### PR TITLE
feat(dss-service): Add FileRevocationSource for cache expiration based on nextUpdate field

### PR DESCRIPTION
**Pull request by gustavoam-asdf**
## Summary

This PR enhances the `RevocationCacheDataLoader` class to implement intelligent cache expiration based on the `nextUpdate` field present in revocation evidence (CRL and OCSP responses). When `nextUpdate` is not available, the system falls back to a configurable default cache expiration time.

#### Smart Cache Based on nextUpdate Field

The class now automatically detects whether it's handling CRL or OCSP requests by checking the DataLoader instance:

- **CommonsDataLoader**: Used for CRL requests  
- **OCSPDataLoader**: Used for OCSP requests

#### NextUpdate Extraction

Added methods to extract the `nextUpdate` field from revocation evidence:

- `extractCRLNextUpdate(byte[] crlData)`: Extracts nextUpdate from CRL data
- `extractOCSPNextUpdate(byte[] ocspData)`: Extracts nextUpdate from OCSP responses

#### Enhanced Expiration Logic

The `isCacheExpired()` method now:

1. **Prioritizes nextUpdate field**: Uses it to determine cache expiration when available
2. **Falls back to defaultCacheExpirationTime**: Uses default expiration when nextUpdate is not present
3. **Configurable via setter**: The `defaultCacheExpirationTime` property can be modified using `setDefaultCacheExpirationTime()`

### CODE REVIEW

**Comment by bsanchezb** [comment]
Hello @gustavoam-asdf ,

Thank you for the PR. 

From what I understand, there is one implementation of the `RevocationCacheDataLoader` class handling both CRL and OCSP responses. How does the implementation have to be used, provided the `CertificateVerifier` class have two separate setters for `CRLSource` and `OCSPSource`?

What about creating a generic abstract implementation of `DSSCacheFileLoader` by extending the [RepositoryRevocationSource](https://github.com/esig/dss/blob/master/dss-spi/src/main/java/eu/europa/esig/dss/spi/x509/revocation/RepositoryRevocationSource.java) class, and providing two implementation one for the `CRLSource` and one for the`OCSPSource`?

KR,
Aleksandr

**Comment by gustavoam-asdf** [comment]
Hey @AleksandrSokolov! You were absolutely right about the architecture. I've completely refactored the `RevocationCacheDataLoader` into separate, cleaner implementations that actually make sense with DSS patterns.

## New Structure

### **FileRevocationSource** (The Base Class)

- Lives in `dss-spi` and extends `RepositoryRevocationSource` like you suggested
- Handles all the common caching logic and smart `nextUpdate` expiration
- Generic enough to work with any revocation type

### **FileCacheCRLSource**

- Specialized for CRL handling in `dss-service/.../crl/`
- Implements `CRLSource` so it fits perfectly in the existing flow

Example:

```java
final var crlCacheDataLoader = new FileCacheCRLSource(cacheDir);

final var dataLoader = new CommonsDataLoader();
final var crlSource = new OnlineCRLSource(dataLoader);

crlCacheDataLoader.setProxySource(crlSource);
```

### **FileCacheOCSPSource**

- Handles OCSP responses in `dss-service/.../ocsp/`
- Implements `OCSPSource` for seamless integration

Example:

```java
final var ocspCacheDataLoader = new FileCacheOCSPSource(cacheDir);

final var dataLoader = new OCSPDataLoader();
final var ocspSource = new OnlineOCSPSource(dataLoader);
ocspSource.setNonceSource(new SecureRandomNonceSource());

ocspCacheDataLoader.setProxySource(ocspSource);
```

## Smart Caching Features

Both implementations include the smart `nextUpdate`-based caching I originally developed:

- **CRL**: Extracts `nextUpdate` from X.509 CRL structure
- **OCSP**: Gets `nextUpdate` from OCSP response `SingleResp`

## Testing

All tests are passing with comprehensive coverage:

- `FileCacheCRLSourceTest` - CRL-specific caching and `nextUpdate` extraction
- `FileCacheOCSPSourceTest` - OCSP response handling and cache logic

**Comment by bsanchezb** [comment]
Hi @gustavoam-asdf ,

Looks better now. If possible, could you please improve the PR with the following:

1) Implement [`DSSCacheFileLoader`](https://github.com/esig/dss/blob/4f3fb4b633a91639e0143b980d2ceea333775d7a/dss-spi/src/main/java/eu/europa/esig/dss/spi/client/http/DSSCacheFileLoader.java) interface by the created `FileRevocationSource` class. This will require to use `DSSDocument`, so try to switch to a `FileDocument` instead of `File`, where possible and makes sense.
2) Try to re-use DSS Utils methods, where possible. For instance, for [file saving](https://github.com/esig/dss/pull/183/files#diff-fde575e27d723e64947476d9a66a04ccc76ed38de2a7656cf96c68c59704c019R205), you may use [DSSUtils#saveToFile](https://github.com/esig/dss/blob/4f3fb4b633a91639e0143b980d2ceea333775d7a/dss-spi/src/main/java/eu/europa/esig/dss/spi/DSSUtils.java#L679) method. Please also see [DSSUtils#getNormalizedString](https://github.com/esig/dss/blob/4f3fb4b633a91639e0143b980d2ceea333775d7a/dss-spi/src/main/java/eu/europa/esig/dss/spi/DSSUtils.java#L695) for URL strings normalizations.
3) Add new classes to the [documentation](https://github.com/esig/dss/blob/4f3fb4b633a91639e0143b980d2ceea333775d7a/dss-cookbook/src/main/asciidoc/_chapters/revocation-data-management.adoc#L4). Maybe separate the "Caching" chapter to two: one for the JDBC source and one for the File Source you created.

Thank you in advance.

I will take a look to the PR in more detail a bit later.

KR,
Aleksandr

**Comment by gustavoam-asdf** [comment]
Hi @bsanchezb,

Thank you for the review and the helpful suggestions.

I will work on points 2 and 3 (re-using `DSSUtils` and updating the documentation).

Regarding point 1, I'd like to better understand the role of the `DSSCacheFileLoader` interface. Could you please clarify the main use case or benefit of implementing it on the `FileRevocationSource` class? Knowing the purpose will help me ensure I implement it correctly.

Thanks again for your guidance.

**Comment by bsanchezb** [comment]
Thank you. I will try to review it during this week or the week after.

KR,
Aleksandr

**Comment by bsanchezb** [comment]
Hi @gustavoam-asdf ,

Thank you again for the PR. It looks pretty good and I'm ready to include it to DSS 6.3.

I added the branch [DSS-3665](https://github.com/esig/dss/tree/DSS-3665) with some changes included:

- Added caching of supplementary information associated with revocation data request, such as original source URL and issuer certificate (for CRL), similarly to a JDBC cache sources;
- The constructors and setters have been modified to align with a design-pattern used within the `FileCacheDataLoader` class;
- Removed the validity checks for retrieved revocation information, as this part is meant to be handled by a `RevocationDataVertifier` class during the validation process;
- Minor improvements to re-use DSS provided Utils classes;
- Updated documentation.

Could you please check the branch if the changes are good with you?

KR,
Aleksandr

**Comment by gustavoam-asdf** [comment]
Hi @bsanchezb,

Thanks for the update. I've checked the [DSS-3665](https://github.com/esig/dss/tree/DSS-3665) branch and your changes look great.

I've also pushed one small modification: I've updated the `saveRevocationToken` method of `FileRevocationSource` to ensure the issuer certificate is stored for all revocation data types, not just for CRLs. This makes the caching behavior more consistent.

Let me know if that works for you.

Thanks again!

**Comment by bsanchezb** [comment]
Hi @gustavoam-asdf ,

There is a reason why original JDBC sources store the issuer only for a CRL. For OCSP responses, the issuer certificate normally is embedded within the OCSP response itself (and unambiguously identified through a ResponderId), while for a CRL the issuer is not known explicitly (we expect it to be a CA certificate, issued the certificate in validation, but the CRL signing can also be delegated to a different certificate).
Therefore, we need to store only the issuer for the CRL, but not for OCSP responses.

Thank you for testing, I will include the PR in DSS 6.3.

Best regards,
Aleksandr

**Comment by gustavoam-asdf** [comment]
Oh, that's right, I forgot that OCSP could have the issuing certificate embedded. You are absolutely right.
Thank you for your attention.

Best regards

**Comment by bsanchezb** [comment]
Merged and included in version 6.3
